### ISSUE TITLE
Fix `shared_ptr` and directors for C# (I describe a possible solution)

### ISSUE THREAD

**Issue by devshgraphicsprogramming** [swig/swig#2410]
First of all I'd like to say that keeping a `shared_ptr` to the original C++ class in the proxy class (and when you use directors, the director) class is a bad idea, let me outline why. _But first I must remark that passing it around everywhere else as a `shared_ptr` is good.

I suppose that many GC languages like C# have some method of keeping an object alive, when unmanaged code such as C++ is holding the only references (for example, a callback). If you ignore the words `C#` and `GCHandle` you should find it applies to many more languages.

# Problem Statement

## You cannot cast back to C# derived class

It would be extremely useful to be able to get working Cross Language Polymorphism by only knowing the C++ pointer.

This could be possible if the C++ director stored a `GCHandle` to the C# wrapper inside itself plus some methods to return its value.
But you simply cannot have that and have the C# wrapper hold a shared_ptr to the C++ director, because thats the definition of a circular reference.

### What happens without shared pointers

Because the C# wrapper class is the "owner" (basically it decides when to call C++`delete`, that is when you call Dispose in C#) of the C++ director if constructed via C# new.

If you don't call Dispose the finalizer will delete the C++ director anyway at some point in the future and probably in the wrong order w.r.t. other references.

### With shared pointers

The C# wrapper class is the "owner" of a single director class that in turn owns a shared pointer to the base class.

When you manually dispose the wrapper or it gets finalized by GC, refcount drops to 0 which is ok.

What is not okay however is that when you pass the smartpointer itself to some other C# or C++ function and lose the reference to the C# wrapper class, if the C# wrapper class was actually derived, "poof" its gone and now you'll get errors upon calling the delegates in the C++ director.

### In general

Without polymorphism its okay to re-create 10000 C# wrapper classes per second or in duplicate for the same underlying C++ wrapper object as long as the `cMemoryOwn` is false or shared pointer is used. However when CLP enters the picture, the C# combined wrapper-and-derived class needs to stay alive for as long as someone is holding the C++ director.

This would also mess up the C# code if it used the C++ interface member variable pointers or getter methods to obtain the  interface and then try to call the virtual methods on that (cause it goes `C#->C# Base class Wrapper->C++ Director->C# Delegate`). 

## How we could keep the C# object alive

We'd like the C# class to stay alive for as long as the C++ shared pointer to the director stays alive, which means doing a `GCHandle.Alloc` in the constructor of the derivations of the C# wrapper class, and then a `Free` somewhere else like a `CppDispose()` method that the C++ director can call in its destructor. 

By applying some logic we can avoid a circular reference, let say that C# holds references to a _shared pointer C# wrapper_ and that  owns a **unique heap allocated** C++ shared pointer which holds a **shared** reference to the director (or a real c++ object which then doesn't hold anything) which is the only C++ object holding a GCHandle to the C# derived class to whom other C# code can hold a reference of.

TL;DR Director holds GC Reference to C# derived class or class wrapper, but not the shared pointer wrapper. Also derived class and class wrapper don't hold any shared pointers.

### Changes needed

1. Don't hold a shared pointer in the C# wrapper class, but also make its finalizer protected
2. require that a `%shared_ptr(T)` inherits from `std::enable_shared_from_this<T>` or has a `shared_from_this()` method
3. Change a the pointer&bool public ctor and add a constructor for just derived classes 
4. if the director has had a `GCHandle` then free it during destruction
5. the `shared_from_this()` in the C# wrapper and any constructor of `shared_ptr` C# wrapper from Type wrapper sets `cMemoryOwn` to 0 on the C# wrapper

### Pointer and Bool public constructor

1. perform a `dynamic_cast` to check if the pointer is actually a director for the type
1.a. if so then attempt to get the `GCHandle` from it, if its not null then its really a C# object
1.a.i don't construct a new C# wrapper class, return the old C# class
2. if the pointer wasn't of the director type then create a new one on the C++ side

### Derived class constructor

1. empty create director then allocate and give the director a `GCHandle` to `this`
2. connect the delegates

### Bonus round

1. just infer `%shared_ptr` from the `std::enable_shared_from_this<T>` inheritance or `shared_from_this()` method
2. since you'd have the GC Handle, you could make the marshalling methods for the delegates `static` taking the C# wrapper GCHandle intPtr value as the first parameter, which would solve #1165 #1579 #1262 and being able to use SWIG with directors in Blazor/Mono-WASM

**Comment by vadz** [comment]
> require that a `%shared_ptr(T)` inherits from `std::enable_shared_from_this<T>` or has a `shared_from_this()` method

I'd like to note that there is plenty of existing code using `%shared_ptr()` with classes not satisfying this requirement and I wouldn't want to break all of it.

**Comment by devshgraphicsprogramming** [comment]
> > require that a `%shared_ptr(T)` inherits from `std::enable_shared_from_this<T>` or has a `shared_from_this()` method
> 
> I'd like to note that there is plenty of existing code using `%shared_ptr()` with classes not satisfying this requirement and I wouldn't want to break all of it.

I understand that, what about the rest of the proposal?

**Comment by vadz** [comment]
Sorry, I didn't really follow through all the details, so I can't say anything useful here. I agree that it would be nice to have cross-language polymorphism working naturally, right now we jump through a lot of hoops to achieve this in our own wrappers. However the approach we use at least works with both C# and Java (and somewhat differently with Python and JS), so in ideal world we'd have something not C#-specific, so I wonder if this proposal could be generalized to at least Java (which is very similar to C#)?

**Comment by devshgraphicsprogramming** [comment]
> Sorry, I didn't really follow through all the details, so I can't say anything useful here. I agree that it would be nice to have cross-language polymorphism working naturally, right now we jump through a lot of hoops to achieve this in our own wrappers. However the approach we use at least works with both C# and Java (and somewhat differently with Python and JS), so in ideal world we'd have something not C#-specific, so I wonder if this proposal could be generalized to at least Java (which is very similar to C#)?

is there a C#-like concept like the GCHandle creation for keeping JS objects alive (prevent GC from collecting) from outside of the JS-world?

**Comment by devshgraphicsprogramming** [comment]
> Do you mean JS or Java? For the latter, you can keep objects alive from JNI, it's actually simpler, or at least more explicit, than with C#. For JS I don't know.

For JS I use Embind, since SWIG can only do bindings for node.js not the browser env.

Anyway because I actually have to finish a project I'm coding up some workarounds which resemble my solution above
```cpp
// All Interfaces must inherit from this type, all C++ must only ever pass these around as an `std::shared_ptr`
// For the reasons why, see https://github.com/swig/swig/issues/2410
class IInterfaceForCS : std::enable_shared_from_this<IInterface>
{
    public:
        // !!IMPORTANT!!! TODO: Inject some code on the C# SWIG side to ensure the C# proxy class has a static constructor which initializes this properly
        // See: https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/static-constructors
        static inline void(const void*) GCHandle_delete = nullptr;

    // none of the non-public members will be exported to Emscripten (which is what we want)
	protected:
        // the derived C++ class must hold the C# GCHandle to its "derived" C# class instance
        // (its really just derived from the C# proxy class)
        IInterfaceForCS(const void* const _csGCHandle) : m_csGCHandle(_csGCHandle)
        {
        }
        // pure virtual destructor to forbid instantiation of the type directly
        virtual ~IInterfaceForCS() = 0;

        // typeless view of the handle we keep around to keep the C# Garbage Collecting the C# implementation's object instance
        const void* const m_csGCHandle;
}

inline IInterfaceForCS::~IInterfaceForCS()
{
    // the function pointer/delegate to our C# static wrapper around `System.Runtime.InteropServices.GCHandle.Free`
    assert(GCHandle_delete);
    GCHandle_delete(m_csGCHandle);
}
```
I think you can see that for JNI there'd barely be a difference, except for the fact that C++ could actually call the handle alloc and free functions itself.

**Comment by devshgraphicsprogramming** [comment]
One more thing, for regular director classes that won't be pointed to by a `shared_ptr` this is obviously a non issue, because there are no factory methods to the C# derived classes on the C++ side, so you can't really create and therefore expect own the C# derived object from the C++ side (we must accept that C++ functions are given the C# objects purely as observers). In this case, there's no expectation of the last reference to a C# derived class being through delegate held onto by C++ created from its method.

Semantically as in plain C++ without smart pointers, when faced with a raw pointer, its the programmer's responsibility to document the ownership and do special somersaults to transfer it, so its okay to expect that one must take care of allocating and managing a GCHandle.

However `shared_ptr`'s semantic is literally the automatic sharing (and by corollary sometimes the transfer) of ownership, so we have a problem there.

**Comment by devshgraphicsprogramming** [comment]
Hmm due to how SWIG works when you do `%shared_ptr` my proposed solution would lead to a circular reference.

But we've added a `GCHandle` helper (like the one you have for string types) to let C++ delete a GC Handle handed to it from C# (for example from a derived object's constructor to C++ proxy class constructor), then user can write themselves some code to tie the lifetime of the C# derived class to the C++ proxy (or some sort of smart pointer) if they so wish.

Will be part of #1262's spiritual successor which will fix the GCHandle memory leak.
